Add support for linear-time mmd estimator.

Srceh opened the pull request:
This PR implements the linear-time estimator in (Lemma14 in [paper](https://www.jmlr.org/papers/volume13/gretton12a/gretton12a.pdf)), as asked in #288.

Srceh committed:
Add support for linear-time MMD estimator.

Srceh committed:
Add support for linear-time MMD estimator.

Srceh committed:
Merge branch 'linear_time_mmd' of github.com:Srceh/alibi-detect into linear_time_mmd

Srceh committed:
Minor fixes.

Srceh commented:
> Do we think users are ever going to have equal reference and test batch sizes in practice? I'd guess almost always the reference set is going to be much larger. I wonder if we'd be better off using the B-stat estimator by default for the linear case rather than Gretton's estimator for equal sample sizes. This additionally has the advantage of a tunable parameter that allows for interpolation between a linear and quadratic time estimator.
> 
> @arnaudvl @Srceh
> 
> Edit: It's actually not quite this simple. However I think we should put some thought into how best to address the n!=m case.

Agree, maybe we can leave the current PR as it is for the linear-time one, and do a separate one for the additional B-stat implementation.

arnaudvl commented:
@Srceh @ojcobb Thinking if it wouldn't be cleaner to have a separate `LinearMMDDrift` and `BMMDDrift` (for lack of a better name) instead of grouping everything in the existing MMD implementation. It would be a bit easier to debug as well and can just share the MMD base class.

Srceh committed:
minor fixes.

Srceh committed:
minor fixes.

Srceh committed:
Merge branch 'linear_time_mmd' of github.com:Srceh/alibi-detect into linear_time_mmd

# Conflicts:
#	alibi_detect/cd/pytorch/mmd.py

Srceh committed:
mnior fixes

Srceh committed:
Created separated class for linear-time estimator

Srceh committed:
Use Gaussian asymptotic distribution under null to calculate test threshold with the linear-time estimator, instead of permutation.

Srceh commented:
Now the linear-time estimator also uses Gaussian under null for the test threshold, so no permutation is required. It should be the fastest at the cost of lower test power and some unused samples.

ojcobb commented on alibi_detect/utils/pytorch/kernels.py:
```diff
@@ -68,16 +68,25 @@ def __init__(
+    def forward(self, x: Union[np.ndarray, torch.Tensor],
+                y: Union[np.ndarray, torch.Tensor],
+                infer_sigma: bool = False,
+                diag: bool = False) -> torch.Tensor:
```
Given that they refer to the same thing, perhaps we could keep consistency between this kwarg name and the naming convention adopted for the squared distance functions? So perhaps `pairwise: bool = True`?

Srceh replied:
Indeed, now uses `pairwise` as suggested.

ojcobb commented on alibi_detect/utils/pytorch/kernels.py:
```diff
@@ -68,16 +68,25 @@ def __init__(
             if self.trainable and infer_sigma:
                 raise ValueError("Gradients cannot be computed w.r.t. an inferred sigma value")
-            sigma = self.init_sigma_fn(x, y, dist)
+            if not diag:
```
Is this a good default behaviour to have? Could we end up with O(n^2) costs in places where the linear time estimator is being used specifically because such a cost would be infeasible?

Srceh replied:
Now directly use the median of the non-pairwise distance.

ojcobb commented on alibi_detect/utils/tensorflow/kernels.py:
```diff
@@ -69,15 +69,24 @@ def __init__(
         return tf.math.exp(self.log_sigma)
 
-    def call(self, x: tf.Tensor, y: tf.Tensor, infer_sigma: bool = False) -> tf.Tensor:
+    def call(self, x: tf.Tensor, y: tf.Tensor,
```
See comments on pytorch version

ojcobb commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,7 +115,43 @@ def batch_compute_kernel_matrix(
 
 
-def mmd2_from_kernel_matrix(kernel_mat: torch.Tensor, m: int, permute: bool = False,
+def linear_mmd2(x: torch.Tensor,
```
Nitpick but probs worth keeping indentation within function definitions consistent with all of the other functions.

Srceh replied:
Fixed, should be consistent all across now.

ojcobb commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -83,7 +109,52 @@ def batch_compute_kernel_matrix(
+    if n != m:
+        raise RuntimeError("Linear-time estimator requires equal size samples")
+    if not permute:
+        k_xx = kernel(x=x[0::2, :], y=x[1::2, :], diag=True)
```
Would this not fail for an odd value of n (and m)? `[0::2, :]`-based subsets would contain extra elements to `[1::2, :]`-based subsets.

ojcobb commented on alibi_detect/cd/mmd.py:
```diff
@@ -18,6 +18,7 @@ def __init__(
             x_ref: Union[np.ndarray, list],
             backend: str = 'tensorflow',
             p_val: float = .05,
+            estimator: str = 'quad',
```
Would `estimator_complexity` be more descriptive? (Or at least make clear in the docstring)

Srceh replied:
Added extra description in the docstring.

ojcobb commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,7 +115,43 @@ def batch_compute_kernel_matrix(
+
+    h = k_xx + k_yy - k_xy - k_yz
+    mmd2 = h.sum() / (n / 2.)
+    var_mmd2 = ((h * h).sum() / (n / 2.)) - (mmd2 ** 2)
```
Should this be sample variance rather than population variance? (i.e. with 1 subtracted from (n/2.))

ojcobb commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,143 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        mmd2 = mmd2.numpy().item()
+        var_mmd2 = var_mmd2.numpy().item()
+        std_mmd2 = np.sqrt(var_mmd2)
+        p_val = 1 - stats.norm.cdf(mmd2 * np.sqrt(n_hat), loc=0., scale=std_mmd2*np.sqrt(2))
```
Nitpick but should this be a t-test?

ojcobb commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,7 +115,43 @@ def batch_compute_kernel_matrix(
+    k_yz = kernel(x=y[0::2, :], y=x[1::2, :], diag=True)
+
+    h = k_xx + k_yy - k_xy - k_yz
+    mmd2 = h.sum() / (n / 2.)
```
Is there a reason we don't just use `h.mean()` and `h.var()`?

Srceh replied:
Now uses `h.mean()`, and `torch.var(, unbiased=True)` in the torch version. TF version uses `tf.reduce_mean` and manual correction.

ojcobb commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -83,7 +109,52 @@ def batch_compute_kernel_matrix(
+        idx = np.random.permutation(m + n)
+        xy = tf.gather(tf.concat([x, y], axis=0), idx)
+        x_hat, y_hat = xy[:n, :], xy[n:, :]
+        k_xx = kernel(x_hat[0::2, :], x_hat[1::2, :], diag=True)
```
Seems like unnecessary duplication

ojcobb commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -83,7 +109,52 @@ def batch_compute_kernel_matrix(
+def linear_mmd2(x: tf.Tensor,
+                y: tf.Tensor,
+                kernel: Callable,
+                permute: bool = False) -> Tuple[tf.Tensor, tf.Tensor]:
```
Is there a reason we offer permute option for tensorflow and not torch?

ojcobb commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -83,7 +109,52 @@ def batch_compute_kernel_matrix(
+def linear_mmd2(x: tf.Tensor,
+                y: tf.Tensor,
+                kernel: Callable,
+                permute: bool = False) -> Tuple[tf.Tensor, tf.Tensor]:
```
Is there a reason we offer permute option for tensorflow and not torch?

Srceh replied:
Legacy issue, now removed for the tensorflow version.

ojcobb commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -83,7 +109,52 @@ def batch_compute_kernel_matrix(
+        idx = np.random.permutation(m + n)
+        xy = tf.gather(tf.concat([x, y], axis=0), idx)
+        x_hat, y_hat = xy[:n, :], xy[n:, :]
+        k_xx = kernel(x_hat[0::2, :], x_hat[1::2, :], diag=True)
+        k_yy = kernel(y_hat[0::2, :], y_hat[1::2, :], diag=True)
+        k_xy = kernel(x_hat[0::2, :], y_hat[1::2, :], diag=True)
+        k_yz = kernel(y_hat[0::2, :], x_hat[1::2, :], diag=True)
```
Seems like unnecessary duplication

Srceh replied:
Removed.

ojcobb commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,7 +115,43 @@ def batch_compute_kernel_matrix(
+
+    h = k_xx + k_yy - k_xy - k_yz
+    mmd2 = h.sum() / (n / 2.)
+    var_mmd2 = ((h * h).sum() / (n / 2.)) - (mmd2 ** 2)
```
Nitpick but should this be sample variance rather than population variance? (i.e. with 1 subtracted from (n/2.))

Srceh replied:
Yes it should be, now both fixed with the unbiased estimation.

Srceh committed:
Fixes according to PR review.

Srceh commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,143 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        mmd2 = mmd2.numpy().item()
+        var_mmd2 = var_mmd2.numpy().item()
+        std_mmd2 = np.sqrt(var_mmd2)
+        p_val = 1 - stats.norm.cdf(mmd2 * np.sqrt(n_hat), loc=0., scale=std_mmd2*np.sqrt(2))
```
Nice spot, now fixed with t-test for both versions.

Srceh committed:
Doc string fixes following review comments.

Srceh commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        Returns
+        -------
+        p-value obtained from the permutation test, the MMD^2 between the reference and test set
+        and the MMD^2 values from the permutation test.
```
Fixed, thanks for spotting.

Srceh commented on alibi_detect/cd/base.py:
```diff
@@ -570,12 +570,16 @@ def predict(self, x: Union[np.ndarray, list], return_p_val: bool = True, return_
+            idx_threshold = int(self.p_val * len(dist_permutations))
+            distance_threshold = np.sort(dist_permutations)[::-1][idx_threshold]
+        else:
+            distance_threshold = tmp_v
```
Yes indeed, guess the best thing to do here is to follow your draft PR's template to modify the linear-time detector.

ascillitoe replied:
I'd say if @arnaudvl and @ojcobb agree with the change in #489 , we should merge that, and then rebase this PR onto the latest `master`.

ascillitoe commented:
@Srceh final thing, were you planning on adding tests and updating the docs?

Srceh commented:
> @Srceh final thing, were you planning on adding tests and updating the docs?

Good question, haven't worked on anything specific, but happy to follow any suggestion.

ascillitoe commented:
> Good question, haven't worked on anything specific, but happy to follow any suggestion.

We'd usually aim to ensure any new functionality added to the library is documented and tested. Since these new detectors come under `MMDDrift`, I think its just a very simple case of making sure the new `estimator` kwarg is documented in `doc/source/methods/mmddrift.ipynb`  🙂 

Could also think about tweaking an existing example to show off (and explain) the linear time MMD if you think if its sufficiently exciting! (this is probs one for @arnaudvl and @ojcobb to weigh in on). I think we mentioned looking at adding some discussion of complexity to the docs at some point, so this could be something to leave until then...

Regarding tests, IMO it might be worth testing the estimators inside `alibi_detect/cd/tensorflow/tests/test_mmd_tf.py` and `alibi_detect/cd/pytorch/tests/test_mmd_pt.py`.

Srceh commented:
> > Good question, haven't worked on anything specific, but happy to follow any suggestion.
> 
> We'd usually aim to ensure any new functionality added to the library is documented and tested. Since these new detectors come under `MMDDrift`, I think its just a very simple case of making sure the new `estimator` kwarg is documented in `doc/source/methods/mmddrift.ipynb` 🙂
> 
> Could also think about tweaking an existing example to show off (and explain) the linear time MMD if you think if its sufficiently exciting! (this is probs one for @arnaudvl and @ojcobb to weigh in on). I think we mentioned looking at adding some discussion of complexity to the docs at some point, so this could be something to leave until then...
> 
> Regarding tests, IMO it might be worth testing the estimators inside `alibi_detect/cd/tensorflow/tests/test_mmd_tf.py` and `alibi_detect/cd/pytorch/tests/test_mmd_pt.py`.

Great thanks! will follow up on these.

arnaudvl commented:
> > > Good question, haven't worked on anything specific, but happy to follow any suggestion.
> > 
> > 
> > We'd usually aim to ensure any new functionality added to the library is documented and tested. Since these new detectors come under `MMDDrift`, I think its just a very simple case of making sure the new `estimator` kwarg is documented in `doc/source/methods/mmddrift.ipynb` slightly_smiling_face
> > Could also think about tweaking an existing example to show off (and explain) the linear time MMD if you think if its sufficiently exciting! (this is probs one for @arnaudvl and @ojcobb to weigh in on). I think we mentioned looking at adding some discussion of complexity to the docs at some point, so this could be something to leave until then...
> > Regarding tests, IMO it might be worth testing the estimators inside `alibi_detect/cd/tensorflow/tests/test_mmd_tf.py` and `alibi_detect/cd/pytorch/tests/test_mmd_pt.py`.
> 
> Great thanks! will follow up on these.

- Yes, tests should always be included for new methods. Can hopefully copy a lot over from existing (e.g. MMD) test and test the linear-time specific changes.
- Same for docs. The method doc can again be quite similar to the MMD one. Similar to online version of MMD.
- For the example, I think it is sufficient to just update an existing example with e.g. 100 runs of the usual MMD detector vs. the linear time one, and show timing differences between those runs (print in cell) but also show that power is likely to go down.

arnaudvl commented:
> > * The method doc can again be quite similar to the MMD one. Similar to online version of MMD.
> 
> Not actually sure we need a new method docs page, since the new detectors are dispatched via the existing `MMDDrift`? Just need to document the `estimator` kwarg in the existing `MMDDrift` doc I think? Could perhaps also add a sentence/paragraph detailing algorithm differences and considerations in choosing one or the other?

Ok yes makes sense.

arnaudvl commented on alibi_detect/cd/mmd.py:
```diff
@@ -4,10 +4,10 @@
 
 if has_pytorch:
-    from alibi_detect.cd.pytorch.mmd import MMDDriftTorch
+    from alibi_detect.cd.pytorch.mmd import MMDDriftTorch, LinearTimeDriftTorch
```
Change to `LinearTimeMMDDriftTorch` for consistency with TensorFlow.

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/mmd.py:
```diff
@@ -40,6 +41,11 @@ def __init__(
+        estimator
+            Estimator used for the MMD^2 computation {'quad', 'linear'}. 'Quad' is the default and
+            uses the quadratic u-statistics on each square kernel matrix. 'Linear' uses the linear
+            time estimator as in Gretton et al. (2014), and the threshold is computed using the Gaussian
```
Would be helpful to add a link to the paper + reference to the specific section in the paper.

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/mmd.py:
```diff
@@ -88,9 +94,21 @@ def __init__(
+            if estimator == 'quad':
+                self._detector = MMDDriftTF(*args, **kwargs)  # type: ignore
+            elif estimator == 'linear':
+                kwargs.pop('n_permutations', None)
```
Best to clarify in the docstrings that `n_permutations` is not used for the *linear* estimator.

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
         return p_val.numpy().item(), mmd2.numpy().item(), mmd2_permuted.numpy()
+
+
+class LinearTimeDriftTorch(BaseMMDDrift):
```
As mentioned before, should probably be `LinearTimeMMDDriftTorch` (what a name...).

ascillitoe replied:
The names just get better and better 😅

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+            data_type: Optional[str] = None
+    ) -> None:
+        """
+        Maximum Mean Discrepancy (MMD) data drift detector using a permutation test, with linear-time estimator.
```
It doesn't use a permutation test though?

Srceh replied:
No it doesn't, removed now.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+            kernel: Callable = GaussianRBF,
+            sigma: Optional[np.ndarray] = None,
+            configure_kernel_from_x_ref: bool = True,
+            n_permutations: int = 100,
```
Should not contain the `n_permutations` kwarg.

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            configure_kernel_from_x_ref=configure_kernel_from_x_ref,
+            n_permutations=n_permutations,
```
Does not need to pass `n_permutations` here.

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+            The kernel evaluation is then averaged over those bandwidths.
+        configure_kernel_from_x_ref
+            Whether to already configure the kernel bandwidth from the reference data.
+        n_permutations
```
Again remove

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+
+    def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        """
+        Compute the p-value resulting from a permutation test using the maximum mean discrepancy
```
Remove reference to permutation test

Srceh replied:
Fixed.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        x_ref, x = self.preprocess(x)
+        n = x.shape[0]
+        m = x_ref.shape[0]
+        n_hat = int(np.floor(min(n, m) / 2) * 2)
```
This behaviour needs to be well documented.

Srceh replied:
Yes, now described in the docstring.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        n = x.shape[0]
+        m = x_ref.shape[0]
+        n_hat = int(np.floor(min(n, m) / 2) * 2)
+        x_ref = torch.from_numpy(x_ref[:n_hat, :]).to(self.device)  # type: ignore[assignment]
```
I believe there is a case to be made that there is an explicit check such that `n == m` and if not, an error is raised. The reason is that silently some unexpected behaviour can occur by only selecting the first `n_hat` reference/test instances. If say the reference data is ordered and contains samples from classes 1,2 and 3, then only choosing `:n_hat` could ignore all samples from class 3 and not form an i.i.d. sample anymore. So my preference would be explicit behaviour around this (raising errors) or if we allow this (which I am not in favour of now) then randomly sample `n_hat` instances from `x_ref` and `x`. Good to have some opinions @jklaise @ascillitoe @ojcobb

ascillitoe replied:
This seems like quite an issue atm. Agree the safest option would be to explicitly check for `n == m`. Otherwise, we could check, and randomly subsample if `n != m`, with a warning raised to inform the user we are doing this.

Srceh replied:
Currently implemented as raise error for `n!=m`. Guess the subsampling should be implemented on a stand-alone part, so that it can be used with other detectors?

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        else:
+            self.k_xx, self.infer_sigma = None, True
+
+    def kernel_matrix(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
```
Method is not used I believe?

Srceh replied:
@arnaudvl The base class requires this method for initialisation, I wonder what would be the preferable solution here? the minimal thing could be to simply leave a pseudo method.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+        # compute kernel matrix for the reference data
+        if self.infer_sigma or isinstance(sigma, torch.Tensor):
+            x = torch.from_numpy(self.x_ref).to(self.device)
+            self.k_xx = self.kernel(x, x, infer_sigma=self.infer_sigma)
```
`self.k_xx` is not used?

Srceh replied:
Fixed, not `self.k_xx` is reused for later calculation.

arnaudvl commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -118,17 +119,144 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, np.ndarray]:
+            mmd2 = mmd2.cpu()
+        mmd2 = mmd2.numpy().item()
+        var_mmd2 = var_mmd2.numpy().item()
+        std_mmd2 = np.sqrt(var_mmd2)
```
Can directly use `torch.std(...)` in `linear_mmd2`? This would remove the few additional lines of code here.

Srceh replied:
The new version uses `np.sqrt(np.clip(var_mmd2, 1e-8, 1e-8))` for numeric stability.

arnaudvl commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -30,6 +30,28 @@ def squared_pairwise_distance(x: torch.Tensor, y: torch.Tensor, a_min: float = 1
     return dist.clamp_min_(a_min)
 
 
+def squared_distance(x: torch.Tensor, y: torch.Tensor, a_min: float = 1e-30) -> torch.Tensor:
```
Can we just apply a reduction to the `squared_pairwise_distance` instead of using an extra function?

Srceh replied:
Now implemented as a single function.

arnaudvl commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,8 +115,48 @@ def batch_compute_kernel_matrix(
+    m = np.shape(y)[0]
+    if n != m:
+        raise RuntimeError("Linear-time estimator requires equal size samples")
+    k_xx = kernel(x=x[0::2, :], y=x[1::2, :], pairwise=False)
```
We should be able to do this at init time (so `self.k_xx` becomes useful again).

Srceh replied:
Fixed, now the kernel matrix is reused for prediction.

arnaudvl commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,8 +115,48 @@ def batch_compute_kernel_matrix(
+    k_xx = kernel(x=x[0::2, :], y=x[1::2, :], pairwise=False)
+    k_yy = kernel(x=y[0::2, :], y=y[1::2, :], pairwise=False)
+    k_xy = kernel(x=x[0::2, :], y=y[1::2, :], pairwise=False)
+    k_yz = kernel(x=y[0::2, :], y=x[1::2, :], pairwise=False)
```
Is `k_yz` the paper notation? B/c it might be easier to follow by just calling it `k_yx`.

Srceh replied:
Typo, thanks for noticing, fixed.

arnaudvl commented on alibi_detect/utils/pytorch/kernels.py:
```diff
@@ -68,16 +68,24 @@ def __init__(
 
-    def forward(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
-                infer_sigma: bool = False) -> torch.Tensor:
+    def forward(self, x: Union[np.ndarray, torch.Tensor],
```
Nitpicking big time here, but let's try to keep same type of indentation as e.g. in the `DeepKernel` below.

arnaudvl commented on alibi_detect/utils/pytorch/kernels.py:
```diff
@@ -68,16 +68,24 @@ def __init__(
 
         x, y = torch.as_tensor(x), torch.as_tensor(y)
-        dist = distance.squared_pairwise_distance(x.flatten(1), y.flatten(1))  # [Nx, Ny]
+        if pairwise:
```
Check my comment in `distance.py` which might make this  `if else` redundant and reduce it to a kwarg of the distance function.

Srceh replied:
Fixed, now as part of the `squared_pairwise_distance` function argument.

arnaudvl commented on alibi_detect/utils/pytorch/kernels.py:
```diff
@@ -68,16 +68,24 @@ def __init__(
+            if pairwise:
+                sigma = self.init_sigma_fn(x, y, dist)
+            else:
+                sigma = (.5 * dist.flatten().sort().values[dist.shape[0] // 2 - 1].unsqueeze(dim=-1)) ** .5
```
Again I think we can avoid the hard-coding of this behaviour and fall back on `self.init_sigma_fn` but with the desired linear detector behaviour.

Srceh replied:
Slightly tricky as the default `init_sigma_fn` is used by other detectors. Might be easier to keep the additional line here?

ascillitoe reviewed (changes requested):
"Requesting changes" to ensure we do not merge until #489 has been merged and `predict` updated in this PR.

Srceh committed:
Added tests and doc for the new estimator, fixes following review comments.

Srceh committed:
fixes on the distance function.

ascillitoe commented on alibi_detect/utils/tensorflow/distance.py:
```diff
@@ -20,14 +26,44 @@ def squared_pairwise_distance(x: tf.Tensor, y: tf.Tensor, a_min: float = 1e-30,
     a_max
         Upper bound to clip distance values.
-
+    pairwise
```
Isn't it a bit unclear to have a function named `squared_pairwise_distance` that optionally computes non-pairwise distances? Perhaps `squared_pairwise_distance` should be renamed `squared_distance`? Or the `pairwise=False` functionality separated out into a separate distance function?

Srceh replied:
I was thinking the same.  It was previously a separate function and @arnaudvl was suggesting making the repeated parts minimal. Guess changing the function name across all related methods would be preferable.

Srceh commented on alibi_detect/utils/pytorch/distance.py:
```diff
@@ -93,8 +115,48 @@ def batch_compute_kernel_matrix(
+    """
+    n = np.shape(x)[0]
+    m = np.shape(y)[0]
+    if n != m:
```
Fixed.

Srceh commented:
> Left a number of comments related to the PyTorch implementation. Let's work through those first and then we can apply the desired changes to TensorFlow as well.

TF version is also fixed for the above ones replied with "fixed".

ascillitoe commented:
@Srceh I've just merged the `score`/`predict` refactoring (#489). This will have introduced a few conflicts you need to resolve. It should simplify your life wrt to the implementation in this PR though!

Srceh commented:
> @Srceh I've just merged the `score`/`predict` refactoring (#489). This will have introduced a few conflicts you need to resolve. It should simplify your life wrt to the implementation in this PR though!

Nice! will start working on that!

Srceh committed:
Following the changes of the score function for mmd detectors (#489).

Merge branch 'master' into linear_time_mmd

# Conflicts:
#	alibi_detect/cd/base.py
#	alibi_detect/cd/mmd.py
#	alibi_detect/cd/pytorch/mmd.py
#	alibi_detect/cd/tensorflow/mmd.py

Commit:
Merge master and resolve conflicts

Commit:
Update tests and pydantic models

ascillitoe commented:
@Srceh I have now merged in the v0.10.0 related changes from `master`. This primarily involved changes to the kwargs related to preprocessing, tweaking some tests, and adding your `estimator` kwarg to the `MMDDrift` pydantic models (see `saving/schemas.py`).

ascillitoe commented on alibi_detect/cd/pytorch/mmd.py:
```diff
@@ -123,21 +124,162 @@ def score(self, x: Union[np.ndarray, list]) -> Tuple[float, float, float]:
         return p_val.numpy().item(), mmd2.numpy().item(), distance_threshold.numpy()
+
+
+class LinearTimeMMDDriftTorch(BaseMMDDrift):
```
Since these new subclasses don't make use of `self.n_permutations` (set in `BaseMMDDrift`), shall we set this to `None`? I had a moment of confusion when updating the tests since `self.n_permuations == 100` when `estimator == 'linear'`.

Srceh replied:
Good point. The default number of permutations then can be initialised in `/cd/mmd.py` when estimator is `'quad'`.

ascillitoe commented on alibi_detect/cd/mmd.py:
```diff
@@ -100,9 +108,21 @@ def __init__(
+                self._detector = MMDDriftTF(*args, **kwargs)  # type: ignore
+            elif estimator == 'linear':
+                kwargs.pop('n_permutations', None)
+                self._detector = LinearTimeMMDDriftTF(*args, **kwargs)  # type: ignore
```
Since the logic to set `self._detector` is located here, we should add additional tests to `alibi_detect/cd/tests/test_mmd.py` to check that the correct subclass is selected conditional on `backend` and `estimator`.

Srceh replied:
Indeed, will modify the tests.

Srceh replied:
Simply rewrite the test to go through different `backend` and `estimator` options, should do the job.

Srceh committed:
Added estimator to mmd tests, changed default permutation to none for base MMD class.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SeldonIO/alibi-detect?pullRequest=475) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SeldonIO/alibi-detect?pullRequest=475) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Ashley Scillitoe<br/>:x: Srceh<br/><hr/>**Ashley Scillitoe** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SeldonIO/alibi-detect?pullRequest=475) it.</sub>
